### DISCUSSION TITLE
bash-tpl requires Bash >= 5.1

### DISCUSSION THREAD

**Discussion by flaix** [TekWizely/bash-tpl#5] [Ideas]
I noticed that `bash-tpl` requires a Bash version 5.1 or higher. That might be worth pointing out in the Readme. Or even a runtime check and warning?

It looks like a really nice tool and people interested in it might stumble over the fact that their machine uses an older version of Bash. One example is macOS, which uses a really old v3 Bash (and will never update). But I also tried it on a Ubuntu based Docker container that wouldn't run `bash-tpl` either, as it comes with a Bash 4.4.

**Comment by TekWizely** [comment]
Greetings !  Thanks for taking the time to report an issue.

Can you clarify what error(s) you're seeing when trying to use it on older bash versions?

I could imagine bash v3.3 (mac) might have some issues, but I _thought_ I tested on v4.4 ...

I'll re-visit my testing but could use some direction if you have any?

Thanks!

-TW

**Comment by flaix** [comment]
Sure.

As for macOS, the version 3.2 does not have _namerefs_ yet. So the use of `local -n` will result in errors of the sort `./bash-tpl: line 331: local: -n: invalid option`. I did a fix for that in flaix/bash-tpl@6e001803ab6fb2d9b6bd39e26e522cf2dd5036bc

The `BASH_REMATCH` variable was read-only until Bash version 5.1, as far as I know. So in older versions writing to it will fail with `bash-tpl: line 311: BASH_REMATCH: readonly variable`. For the respective code lines you may check flaix/bash-tpl@2de4f6ed96e9622540b2d0f5316a4af72b8121f5

I changed the code in my fork to make it work under macOS. Which should also make it work on a Github ubuntu-18.04 runner, I guess. I am not sure what your policy is regarding supporting older Bash versions, so I did not create pull requests. I would assume that you may be targeting Bash 5 and above. Hence the remark if it may make sense to mention that in the Readme.

Also, I really just made it work for me, as I was actually trying to use it to solve another problem. :wink: As it usually is the case. But if you are interested, in the changes, you can find them in [flaix/bash-tpl](https://github.com/flaix/bash-tpl) in the branches `no-nameref` and `readonly-bash-rematch`. Feel free to create pull requests from them or let me know if I should do so. Of course I am also interested if you have a better or cleaner solution than I did. 😄

**Comment by flaix** [comment]
Hey, you fixed it in your code. That is great, I can convert from my fork to your version then, once merged. 👍

I see you opted for a `printf -v` instead of using an `eval`. I wasn't aware of `printf -v`. Is that faster or safer, or what makes it the better choice for this case?

**Reply by TekWizely** [reply]
Hey again !

Q: Could you try downloading the script from the PR and testing it in your system ?  It would be nice to get a secondary confirmation of the changes actually working :)  Feel free to comment on the PR with your findings.

re: `printf -v`

I _think_ it exists as a convenience for creating a variable from a complicated string format.

**BUT** unlike `declare`, it plays nicer with variable scoping. ie. `declare` in a function makes the variable local, ignoring any already-present variable in scope.

Also `printf` is a bash builtin and so is very efficient.

Side Note: You should consider `printf` over `echo` within your scripts in almost ALL cases.

re: better than `eval`

Eval is considered both slow (has to create a sub-shell and re-parse the content) and risky (less safe), and is generally frowned upon by the community.  As such, it should always be a very last resort.

Thanks for participating and please do try out the version from the PR if you get a chance.

-TW

**Comment by flaix** [comment]
I can say that it works well now. I switched to your version this morning and have been using it successfully during the day already.
Thanks for the prompt reaction. It is nice to be able to use the normal upstream.

Really a nifty little tool, I must say!

**Comment by TekWizely** [comment]
> I am happily generating Dockerfiles from a template now. 👍

Side note: i would love to see what your Dockerfile template looks like.

I'm trying to figure out what a good `killer app` would be for bash-tpl and then put it in the README so ppl can get a sense for just how powerful it really is.

Its hard to convey to people that bash-tpl unlocks the full power of their favorite shell scripting language inside a text document !

Anyway, thanks a lot for your assistance and for using bash-tpl !

**Comment by flaix** [comment]
Well, you have a show-and-tell topic, probably for that reason. Would you like me to create a post?
I guess it would be better to create a smaller example template, as my actual Dockerfiles template is rather long, while not really complicated.

**Reply by TekWizely** [reply]
A smaller example might be nice to act as a starter for others.

Q: Do you blog?  If so you could consider a blog post on whatever platform you use.  It might get more eybeballs than posting in the discussion section here.

SO yesterday I added a [workflow action](https://github.com/TekWizely/bash-tpl/blob/main/.github/workflows/bats-multi-bash.yml) to run the BATS tests against the 4 bash versions.

As part of researching how to set it up, I discovered that the official Bash docker images are built using jq and gawk against a JSON configuration for the version spec and a Docker.template file:

* https://github.com/tianon/docker-bash

I think the gawk template language is pretty ugly and suspect your Dockerfile template will look much nicer !

**Reply by flaix** [reply]
I never took up blogging. Always felt I should, as it was the thing to do back in the days, but I never got myself motivated enough to really start.

Yeah, I know about the jqt-gawk thing. All the "official" Docker images, i.e. from the Docker community builds, are built this way. I understand why, but I don't have a huge matrix of different builds and it looked too complicated for what I wanted to do. So I went looking for a smaller solution, preferably just a bash script. That is how I came across bash-tpl, among many other such tools, and then settled for bash-tpl in the end.

I don't know about nice though. 😆 They have less whitespace, I must admit.

**Reply by TekWizely** [reply]
I was not aware of the general jqt-gawk thing, TIL !

I hear you on the blogging.  If you haven't seen https://dev.to/ yet you might take a look - they make it very easy to create blog entries (all markdown based with tags, etc)

Regardless, I'd be thrilled to have a post here on the show-and-tell topic too !

**Reply by TekWizely** [reply]
fyi here's my dev.to account: https://dev.to/tekwizely
